### ISSUE TITLE
Text-level diagnostics needed

### ISSUE THREAD

**Issue by martindholmes** [projectLEMDO/lemdoIssues#3]
This is a list of nice-to-have features for a text-level diagnostic process, gathered from editors:

Please add any diagnostic requests here so we can develop feature requests for our developer. 

**Working Diagnostics** 

- Can we check the accuracy of quotations in the critical paratexts (by checking against their own modern texts)?  
- Flag links to non-published documents 
- Number of words spoken by each character 
- Number of speeches by each character 
- Characters who speak with the longest words/shortest words 
- Number of words in text 
- Number of speeches in text 
- Something that could show when speakers are onstage? We could see who appears on stage together? --> similar to B G-H's – to do this, we'll have to add a @who attribute to entrances and exits -- ask LE for tips on how to encode this. 

**Text Analysis** 

From B G-H: 

- Generate cue-scripts by character  
- Generate dialogue by character (minus cues)  
- Generate a table of entrances and exits by scene (i.e., the building blocks of a doubling chart) -- similar to LE's request to see who appears on stage together.

**Comment by martindholmes** [comment]
Some work has already been completed in code/diagnostics/diagnostics_text.xsl.

**Comment by martindholmes** [comment]
I've added the current text-level diagnostics processing as the default transformation for LEMDO XML files.

**Comment by martindholmes** [comment]
In rev 16313, I added a listing of entrances and exits.

**Comment by martindholmes** [comment]
In rev 17367, I added proper handling for milestone act and scene divisions for entrances and exits.

**Comment by martindholmes** [comment]
@JanelleJenstad @LEMDO-PM I need to get your input on this ticket at this stage. Some of the remaining items (such as tables of entrances and exits) cannot be done with additional encoding which doesn't yet exist (such as stage/@who), while others such as cue scripts are going to be massive -- the diagnostics would be longer than the play itself -- so I think they should be a separate feature request handled in a different process. Links to non-published documents will not typically appear in play texts, they'll be in ancillary files, so this suggests that this diagnostic is not just for play texts but for any files; is that right? Would that be better handled by a diagnostic across the whole collection?

**Comment by martindholmes** [comment]
Over the vacation I added diagnostics for unused anchors and pointers to non-existent anchors; this is crucial to prevent committing of files which are not invalid, but which will result in build breaks.

**Comment by martindholmes** [comment]
@JanelleJenstad @LEMDO-PM Ping on this -- any response?

**Comment by LEMDO-PM** [comment]
I'll leave the April comment to Janelle who knows more about what we want for the long-term LEMDO goals. Re: unused anchors and pointers, I think they're great and already coming in handy :)

We do want to be careful with the unused anchors diagnostic--it only looks within a single edition, right? And we might have people linking to an anchor in one edition from a different edition (i.e., some anchors may be being used even if we can't see that within the edition itself).

For myself and Janelle, we'll need to start teaching people to use those diagnostics. I think most people don't know they exist, so really don't know how to use them. It'll be a combo of writing documentation and incorporating it into training.

**Comment by martindholmes** [comment]
@LEMDO-PM Yes, an unused anchor is not necessarily an error; it may be there because you plan to use it, or it may be pointed-at by a text somewhere completely different (although what would be done with that at rendering time is a mystery). So deciding whether to delete an unused anchor would depend on a thorough search of the whole repository.

**Comment by martindholmes** [comment]
I think the who's-on-stage, cue-scripts and dialog lists per character should be a separate ticket, since they're not diagnostic. If no-one disagrees, I'll close this ticket -- I have one other one open on checking the CSS as part of the same process, but that may not be practical; meanwhile, everything truly diagnostic is done.
